### ISSUE TITLE
Windows uses?

### ISSUE THREAD

**Issue by swails** [choderalab/ambermini#33]
I was wondering what the use-case is here for supporting Windows?  Is anybody actively using WIndows versions of ambermini?

**Comment by rmcgibbo** [comment]
Of the developers watching this repo (you, john, patrick, and I), I believe that there are no Windows users. But on https://anaconda.org/omnia/ambermini/files?version=15.0.4, the Windows build for omnia's anaconda channel, there are about the same number of downloads of this package for Windows as there are for OS X.

I've always personally tried to push for more Windows support (IIRC I submitted the required patches to these tools) because I just think it's the proper professional thing to do.

**Comment by rmcgibbo** [comment]
> Of the developers watching this repo (you, john, patrick, and I), I believe that there are no Windows users.

Obviously not all of the users are commenting on the github repo in general though, especially if they're only directly use a downstream.

**Comment by swails** [comment]
> Obviously not all of the users are commenting on the github repo in general though, especially if they're only directly use a downstream.

Right, but I was wondering if there were immediate developer use-cases for this support.  I agree that improved Windows support is the more professional thing to do, but practical time restrictions are quite prohibitive in terms of broadly enabling Windows support in the whole Amber code base.  We've made good strides, but it's far from finished.  We're mainly trying to gauge how important of a priority this should be.  (And yes, you submitted Windows fixes to this repo, which I moved over to the main Amber repo in several instances).

**Comment by rmcgibbo** [comment]
Are we talking about this repo, or the rest of the Amber codebase?

**Comment by swails** [comment]
For all of Amber we wanted to see where (and how) the windows-compatible bits of Amber out there are being used to help us prioritize what we should do with Windows and the rest of Amber.

**Comment by rmcgibbo** [comment]
I see. I'll adjust my comments accordingly. Here are a couple

- A *lot* of people use Windows. I think this makes it an important target, I think, especially for system setup, and analysis.
- I also think it's an important target for teaching and tutorials.
- The reason I personally worked on the Windows support for cpptraj is that I wanted to add some tests to the mdtraj CI setup that would test some functions against cpptraj. cross-platform testing is an important part of mdtraj, so that was only possible if cpptraj ran on Windows.
   * So if you consider enabling other projects' cross-platform CI as an immediate developer use case, then I guess there's one.
- Especially with a mingw toolchain (or if you switch to CMake), it's really not very hard, IMO to build software on Windows.
- In the interests of comity and decorum, I will refrain from commenting on the existing Amber build system.

**Comment by swails** [comment]
> In the interests of comity and decorum, I will refrain from commenting on the existing Amber build system.

haha... it is what it is :).  If you ask Amber devs about it, you'll repeatedly hear things like "it's not great, but it's a lot better than it once was".  Amber is a very old program, and a lot of the tools that are natural to use now simply didn't exist (or were *highly* unreliable if they did exist) when the current build system was constructed.

Moving to something more modern has a lot of barriers:

- Far more core Amber devs understand the current build system enough to modify and improve it.  Far fewer (if anybody) knows CMake well enough to migrate to that (and enough people detest autotools enough to not consider that route).  Since Amber relies heavily on temporary contributors, moving from a system many people understand to one that only a single post-doc or grad student does is inherently risky (what happens when they move on?)
- It would be time-consuming to migrate to CMake, and any estimate of that time requirement will almost certainly be an underestimate.
- Nobody knows all of the codes well enough to do this for all components without significant time investments.

The idea of moving to CMake was floated before, but nobody is going to put in that kind of time for something they're not sure will be accepted (even if it does get "finished").

**Comment by rmcgibbo** [comment]
Yes, I understand. If I personally worked on Amber, I would strongly advocate changing the build system, but I don't; It's not my battle to fight.

Anyways, that's really a separate thing from Windows.

**Comment by swails** [comment]
> If I personally worked on Amber, I would strongly advocate changing the build system,

Many people have before, even in recent memory.  It's not something "outsiders" would have seen, obviously, but it has happened.  It's not a battle you would win as a more-than-likely-temporary contributor, though...  FWIW, I supported the guy who originally wanted to move to CMake, but I didn't know nearly enough about it at the time (or now, really) to help much.  So it was dropped.

**Comment by rmcgibbo** [comment]
That's fine. As I said, it's not my battle to fight.

**Comment by hainm** [comment]
> But on https://anaconda.org/omnia/ambermini/files?version=15.0.4, the Windows build for omnia's anaconda channel, there are about the same number of downloads of this package for Windows as there are for OS X.

max downloads:
```python
linux: 2221
oxs: 48
win: 53
```

I am not sure if win and osx numbers are significant. Probably people are working on Linux cluster rather personal computers.

Is there any chance that other packages in `omnia` pull its build (ambermini) for both osx and win? If so, the downloads might no reflect the win users.

**Comment by rmcgibbo** [comment]
It's not clear to me that these numbers are meaningful at all. Most of the downloads are probably from Travis-CI anyways.

**Comment by hainm** [comment]
uhm,

can you explain your point in this (2nd sentence).

> Of the developers watching this repo (you, john, patrick, and I), I believe that there are no Windows users. But on https://anaconda.org/omnia/ambermini/files?version=15.0.4, the Windows build for omnia's anaconda channel, there are about the same number of downloads of this package for Windows as there are for OS X.

**Comment by rmcgibbo** [comment]
My point was that Windows and OS X appear to have the same roughly number of downloads. Linux obviously has many more, most of which are probably due to Travis-CI. Number of users != number of downloads.

I also don't think that these numbers are particularly important.

**Comment by hainm** [comment]
Sorry, I probably misunderstood you.

I though many people use OSX and when you said about the number of downloads in Win and OSX are similar, I thought you meant there were significant number of Win users.
